### PR TITLE
started to add the tmp variable is in/out variable to single and doub…

### PR DESCRIPTION
**Pull request by eldadHaber**
Made changes to single and doubleSymLayer to support tmp as input and output
Changed the ResNN to support this change

This should propagate up ...

### CODE REVIEW

File: src/integrators/ResNN.jl [hunk 1]
```diff
@@ -61,7 +67,7 @@ function  apply(this::ResNN{T},theta::Array{T},Y0::Array{T},doDerivative=true) w
 
     Ydata = zeros(T,0,nex)
     for i=1:this.nt
-        Z,dummy,tmp[i,2] = apply(this.layer,theta[:,i],Y,doDerivative)
+        Z,dummy,tmp[i,2] = apply(this.layer,theta[:,i],Y,doDerivative,tmp[i,2])
```
Comment (klensink) [inline src/integrators/ResNN.jl:70]: tmp[i, 2] creates a sub array that copies the data out of tmp into a temporary variable. I suspect that this means you would not really be saving many memory allocation passing the input like this. 

I suspect would need to either pass `view(tmp, i, 2)` to create a reference to that section in memory, or pass `tmp, i, 2' as three input arguments and modify tmp in place within the function's scope to accomplish this with no allocations.

**Comment by eldadHaber** [comment]
Good points

Can you do the change?
We can discuss after how to extend it. 

E

**Comment by klensink** [comment]
Yup, I'll test it out tomorrow and accept the PR once I get it working.

File: src/integrators/ResNN.jl [hunk 2]
```diff
@@ -53,6 +53,12 @@ function  apply(this::ResNN{T},theta::Array{T},Y0::Array{T},doDerivative=true) w
     nex = div(length(Y0),nFeatIn(this))
     Y   = reshape(Y0,:,nex)
     tmp = Array{Any}(this.nt+1,2)
```
Comment (klensink) [inline src/integrators/ResNN.jl:55]: @eldadHaber @lruthotto 

Does `tmp` need to be an `Array{Any}`? It looks like `tmp` could be an `Array{Array{T}, 2}`

**Comment by eldadHaber** [comment]
Not sure if Array{Array{T},1} or Array{Array{T},2}

**Comment by klensink** [comment]
From how you defined it here (`tmp = Array{Any}(this.nt+1,2)`) it looks like it would be an `Array{Array{T}, 2}`. I tried to quickly edit my comment above, but it looks like you got this by email so you saw my typo. 

I was more looking to confirm if the elements of `tmp` will be arrays?

**Comment by eldadHaber** [comment]
In general yes

**Comment by lruthotto** [comment]
I'm skeptical about making tmp an `Array{Array{T}, 2}`. Note that in the second column, we'll have temp variables from the layers. For some layers, this will itself be an `Array{T,2}`. So, on the ResNN level, I don't think we can constrain the type.

**Comment by lruthotto** [comment]
Might be worth trying this though. I'd expect the tests to fail if you add the constraint.

**Comment by DavidBegert** [comment]
closing due to outdated

**Pull request closed without merging**